fortanix-vme/nsm/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+    pub max_pcrs: u16,
+    /// The PCRs that are read-only.
+    pub locked_pcrs: BTreeSet<u16>,
+    /// The digest of the PCR Bank
+    pub digest: Digest,
+}
+
+impl Description {
+    pub fn version_major(&self) -> u16 {
+        self.version_major
+    }
+
+    /// Minor API changes are denoted by `minor_version`. Minor versions should be backwards compatible.
+    pub fn version_minor(&self) -> u16 {
+        self.version_minor
+    }
+
+    /// Patch version. These are security and stability updates and do not affect API.
+    pub fn version_patch(&self) -> u16 {
+        self.version_patch
+    }
+
+    /// `module_id` is an identifier for a singular NitroSecureModule
+    pub fn module_id(&self) -> &String {
+        &self.module_id
+    }
+
+    /// The maximum number of PCRs exposed by the NitroSecureModule.
+    pub fn max_pcrs(&self) -> u16 {
+        self.max_pcrs
+    }
+
+    /// The PCRs that are read-only.
+    pub fn locked_pcrs(&self) -> &BTreeSet<u16> {
+        &self.locked_pcrs
+    }
+
+    /// The digest of the PCR Bank
+    pub fn digest(&self) -> Digest {
+        self.digest
+    }
+}
+
+impl TryFrom<Response> for Description {
+    type Error = Error;
+
+    fn try_from(response: Response) -> Result<Self, Self::Error> {
+        match response {
+            Response::DescribeNSM {
+                version_major,
+                version_minor,
+                version_patch,
+                module_id,
+                max_pcrs,
+                locked_pcrs,
+                digest,
+            } => Ok(Description {
+                    version_major,
+                    version_minor,
+                    version_patch,
+                    module_id,
+                    max_pcrs,
+                    locked_pcrs,
+                    digest,
+                  }),
+            Response::Error(code) => Err(code.into()),
+            _ => Err(Error::InvalidResponse),
+        }
+    }
+}
+
+impl Nsm {
+    pub fn new() -> Result<Self, Error> {
+        let fd = nsm_driver::nsm_init();
+        if fd < 0 {
+            Err(Error::CannotOpenDriver)
+        } else {
+            Ok(Nsm(fd))
+        }
+    }
+
+    pub fn attest(&mut self, user_data: Option<ByteBuf>, nonce: Option<ByteBuf>, public_key: Option<ByteBuf>) -> Result<AttestationDocument<Unverified>, Error> {
+        let req = Request::Attestation {
+            user_data,
+            nonce,
+            public_key,
+        };
+        match nsm_driver::nsm_process_request(self.0, req) {
+            Response::Attestation { document } => Ok(AttestationDocument::from_slice::<Mbedtls>(document.as_slice())?),
+            Response::Error(code) => Err(code.into()),
+            _ => Err(Error::InvalidResponse),
+        }
+    }
+
+    pub fn describe_pcr(&mut self, idx_pcr: u16) -> Result<Pcr, Error> {
+        let req = Request::DescribePCR {
+            index: idx_pcr,
+        };
+        nsm_driver::nsm_process_request(self.0, req).try_into()
```

Comment:
A disadvantage of building a `Pcr` instance from a `Response` like this, is that we do not detect the (admittedly, bizarre) case when the driver would respond with a `ExtendPCR` response to a `DescribePCR` request or vice versa. I think I personally wouldn't do the conversion from `Response`s to `Pcr`s or `Description`s using the `TryFrom` trait.

Reply:
It's an edge case, but may be worth while to avoid it.